File: src/main/scala/org/globalmoney/mandrill/MandrillPlugin.java
package org.globalmoney.mandrill

import java.io.{File, FilterOutputStream, PrintStream}
import javax.mail.internet.InternetAddress

import org.apache.commons.mail._
import play.api._
import scala.collection.JavaConverters._

class MailerPlugin(app: play.api.Application) extends play.api.Plugin {

  private lazy val mock = app.configuration.getBoolean("smtp.mock").getOrElse(false)

  private lazy val mailerInstance: MailerAPI = {
    if (mock) {
      MockMailer
    } else {
      val smtpHost = app.configuration.getString("smtp.host").getOrElse(throw new RuntimeException("smtp.host needs to be set in application.conf in order to use this plugin (or set smtp.mock to true)"))
      val smtpPort = app.configuration.getInt("smtp.port").getOrElse(25)
      val smtpSsl = app.configuration.getBoolean("smtp.ssl").getOrElse(false)
      val smtpTls = app.configuration.getBoolean("smtp.tls").getOrElse(false)
      val smtpUser = app.configuration.getString("smtp.user")
      val smtpPassword = app.configuration.getString("smtp.password")
      val debugMode = app.configuration.getBoolean("smtp.debug").getOrElse(false)
      val smtpTimeout = app.configuration.getInt("smtp.timeout")
      val smtpConnectionTimeout = app.configuration.getInt("smtp.connectiontimeout")
      new CommonsMailer(smtpHost, smtpPort, smtpSsl, smtpTls, smtpUser, smtpPassword, debugMode, smtpTimeout, smtpConnectionTimeout) {
        override def send(email: MultiPartEmail): String = email.send()
        override def createMultiPartEmail(): MultiPartEmail = new MultiPartEmail()
        override def createHtmlEmail(): HtmlEmail = new HtmlEmail()
      }
    }
  }

  override lazy val enabled = !app.configuration.getString("apachecommonsmailerplugin").filter(_ == "disabled").isDefined

  override def onStart() {
    mailerInstance
  }

  def instance = mailerInstance
}

File: src/main/scala/org/globalmoney/mandrill/MandrillPlugin.scala
package org.globalmoney.mandrill

import java.util.Date

import com.typesafe.config.ConfigRenderOptions
import play.api.Logger
import play.api.libs.json.{JsError, JsSuccess, Json}
import play.api.Application
import play.api.libs.ws._
import scala.concurrent.Future
import play.api.libs.concurrent.Execution.Implicits._
import Serializer._
class MandrillPlugin(app: Application) extends play.api.Plugin {

  private lazy val mock = app.configuration.getBoolean("mandrill.mock").getOrElse(false)

  private lazy val mandrillInstance: MandrillAPI = {
    val defaultMailRequestConf = app.configuration.getObject("mandrill.mail")
    val defaultMail = if (defaultMailRequestConf.isDefined) {
      Json.parse(defaultMailRequestConf.get.render(ConfigRenderOptions.concise())).validate[Email] match {
        case mail: JsSuccess[Email] =>
          mail.get
        case e: JsError =>
          throw new RuntimeException("Could not parse default mail request (mandrill.default.mail.json in application.conf). Error:" + e.errors.toString())
      }
    } else {
      throw new RuntimeException("Default mail request not found (mandrill.mail in application.conf)")
    }

    if (mock) {
      new MockMandrill(defaultMail)
    } else {
      val apiKeyStr = app.configuration.getString("mandrill.key").getOrElse(
          throw new RuntimeException("mandrill.key needs to be set in application.conf in order to use this plugin (or set mandrill.mock to true)")
      )

      val apiUrlStr = app.configuration.getString("mandrill.url").getOrElse("https://mandrillapp.com/api/1.0")
      val apiMailUrlStr = app.configuration.getString("mandrill.mail.url").getOrElse("/messages/send.json")

      new RestMandrill(apiUrlStr, apiMailUrlStr, apiKeyStr, defaultMail, app)
    }
  }

  override lazy val enabled = true

  override def onStart() {
    mandrillInstance
  }

  def instance = mandrillInstance
}


object MandrillPlugin {

  def send(email: Email, async: Option[Boolean] = Some(false), ipPool: Option[String] = None, sendAt: Option[Date] = None)(implicit app: play.api.Application) =
    app.plugin(classOf[MandrillPlugin]).get.instance.send(email, async, ipPool, sendAt)

  def sendTemplate(template: Template, email: Email, async: Option[Boolean] = Some(false), ipPool: Option[String] = None, sendAt: Option[Date] = None)(implicit app: play.api.Application) =
    app.plugin(classOf[MandrillPlugin]).get.instance.sendTemplate(template, email, async, ipPool, sendAt)
}


trait MandrillAPI {
  def send(email: Email, async: Option[Boolean] = Some(false), ipPool: Option[String] = None, sendAt: Option[Date] = None): Future[List[Result]]
  def sendTemplate(template: Template, email: Email, async: Option[Boolean] = Some(false), ipPool: Option[String] = None, sendAt: Option[Date] = None): Future[List[Result]]
}

class MockMandrill(defaultMail: Email) extends MandrillAPI {
  override def send(email: Email, async: Option[Boolean] = Some(false), ipPool: Option[String] = None, sendAt: Option[Date] = None): Future[List[Result]] = {
    sendRequest(None, email, async, ipPool, sendAt)
  }

  override def sendTemplate(template: Template, email: Email, async: Option[Boolean] = Some(false), ipPool: Option[String] = None, sendAt: Option[Date] = None): Future[List[Result]] = {
    sendRequest(Some(template), email, async, ipPool, sendAt)
  }

  def sendRequest(template: Option[Template], email: Email, async: Option[Boolean] = Some(false), ipPool: Option[String] = None, sendAt: Option[Date] = None): Future[List[Result]] = {
    Logger.debug(s"Mock mandrill email send: ${Serializer.printJson(Request("", email.withDefaults(defaultMail), template), pretty = true)}")
    Future(List[Result]())
  }
}

class RestMandrill(apiUrl: String, apiMailUrl: String, key: String, defaultMail: Email, app: Application) extends MandrillAPI {
  val mailUrl = apiUrl.concat(apiMailUrl)
  implicit val application = app

  override def send(email: Email,
                    async: Option[Boolean] = Some(false),
                    ipPool: Option[String] = None,
                    sendAt: Option[Date] = None): Future[List[Result]] = {
    sendRequest(None, email, async, ipPool, sendAt)
  }

  override def sendTemplate(template: Template, email: Email, async: Option[Boolean], ipPool: Option[String], sendAt: Option[Date]): Future[List[Result]] = {
    sendRequest(Some(template), email, async, ipPool, sendAt)
  }

  def sendRequest(template: Option[Template],
           email: Email,
           async: Option[Boolean] = Some(false),
           ipPool: Option[String] = None,
           sendAt: Option[Date] = None): Future[List[Result]] = {

    val request = Request(key, email.withDefaults(defaultMail), template, async, ipPool, sendAt)
    Logger.debug(s"Server request: ${Serializer.printJson(request, pretty = true)}")

    val responseFuture = WS.url(mailUrl).post(Serializer.printJson(request))

    responseFuture map { resp =>
      Logger.debug(s"Server response ${resp.status}, body: ${resp.body}")

      resp.status match {
        case 200 =>
          resp.json.as[List[Result]]
        case _ =>
          throw new MandrillSendingException(resp.json.as[MandrillError])
      }
    }
  }
}

